src/transports/smart_protocol.c

```diff
@@ -298,7 +301,21 @@ static int fetch_setup_walk(git_revwalk **out, git_repository *repo)
 		if (git_reference_type(ref) == GIT_REF_SYMBOLIC)
 			continue;
 
-		if ((error = git_revwalk_push(walk, git_reference_target(ref))) < 0)
+                /* only walk over commits */
+                oid = git_reference_target(ref);
+                if ((error = git_object_lookup(&oobj, repo, oid, GIT_OBJ_ANY)) < 0)
+                        goto on_error;
+
+                error = git_object_peel(&obj, oobj, GIT_OBJ_COMMIT);
```

Comment:
Can you explain why are you peeling here?  What type of objects could you find above (in the lookup) that would need to be peeled?  It seems like this could be simplified as:

    oid = git_reference_target(ref);
    if ((error = git_object_lookup(&obj, repo, oid, GIT_OBJ_ANY)) < 0)
        goto on_error;

    if (git_object_type(obj) != GIT_OBJ_COMMIT) {
        git_object_free(obj);
        continue;
    }

It might be nicer to specify the type to `git_object_lookup` and modify `git_object_lookup` to return a unique error code when the type of the object is not the type requested.

Reply:
In my use-case the first object is actually an annotated tag, which when peeled points to a blob. But in other cases it could point to a commit, and in this case the commit can be added to the revwalk.

Reply:
Got it, thanks for the explanation.